Sources/SwiftParser/Attributes.swift

```diff
@@ -918,6 +939,75 @@ extension Parser {
   }
 }
 
+extension Parser {
+  mutating func parseABIAttributeArguments(missing: Bool = false) -> RawABIAttributeArgumentsSyntax {
+    let providerDecl: RawABIAttributeArgumentsSyntax.Provider
+    if missing || at(.rightParen) {
+      // There are two situations where the left paren might be missing:
+      //   1. The user just forgot the paren: `@abi var x_abi: Int) var x: Int`
+      //   2. The user forgot the whole argument list: `@abi var x: Int`
+      // Normally, we would distinguish these by seeing if whatever comes after
+      // the attribute parses as an argument. However, for @abi the argument is
+      // a decl, so in #2, we would consume the decl the attribute is attached
+      // to! This leads to a lousy diagnostic in that situation.
+      // Avoid this problem by simply returning a missing decl immediately.
+      // FIXME: Could we look ahead to find an unbalanced parenthesis?
+      providerDecl = .unsupported(
+        RawDeclSyntax(
+          RawMissingDeclSyntax(
+            attributes: self.emptyCollection(RawAttributeListSyntax.self),
+            modifiers: self.emptyCollection(RawDeclModifierListSyntax.self),
+            arena: arena
+          )
+        )
+      )
+    } else {
+      switch nextDeclarationKeyword() {
+      case .group?:
+        providerDecl = RawABIAttributeArgumentsSyntax.Provider(parseDeclarationGroupHeader())!
```

Comment:
I don’t like the use of the force unwrap here and thus the lack of type system checking that `RawABIAttributeArgumentsSyntax.Provider` can actually represent the parsed declaration header.

I don’t think the force unwrap is needed if you created the `RawABIAttributeArgumentsSyntax.Provider.declGroup` case explicitly.